In [340]:
import pandas as pd
import numpy as np

Inventories dataframe

In [341]:
df_inventories = pd.concat(
    map(pd.read_csv, ['data/lego_sets_2012.csv', 'data/lego_sets_2013.csv', 'data/lego_sets_2014.csv', 'data/lego_sets_2015.csv', 'data/lego_sets_2016.csv', 'data/lego_sets_2017.csv', 'data/lego_sets_2018.csv', 'data/lego_sets_2019.csv', 'data/lego_sets_2020.csv', 'data/lego_sets_2021.csv', 'data/lego_sets_2022.csv']), ignore_index= True)
df_inventories

Number          Theme            Subtheme  Year  \
0                  3061-1        Friends                 NaN  2012   
1                  3063-1        Friends                 NaN  2012   
2                  3065-1        Friends                 NaN  2012   
3                  3183-1        Friends                 NaN  2012   
4                  3184-1        Friends                 NaN  2012   
...                   ...            ...                 ...   ...   
8955  ISBN9781797214139-1          Books  Dorling Kindersley  2022   
8956  ISBN9781797217604-1          Books           Chronicle  2022   
8957           REDBRICK-1    Promotional   LEGO Brand Stores  2022   
8958         SUMMERWAVE-1    Promotional   LEGO Brand Stores  2022   
8959               YOTT-1  Miscellaneous                 NaN  2022   

                                               Set name  Minifigs  Pieces  \
0                                        City Park Cafe       2.0   222.0   
1                                 Heartlake Flying Club       1.0   195.0   
2                                   Olivia's Tree House       1.0   191.0   
3                          Stephanie's Cool Convertible       1.0   130.0   
4                                      Adventure Camper       2.0   309.0   
...                                                 ...       ...     ...   
8955                                    Build Every Day       NaN     NaN   
8956  LEGO in Focus: Explore the Miniature World of ...       NaN     NaN   
8957                          Buildable 2 x 2 Red Brick       NaN    66.0   
8958                                        Summer Wave       NaN    39.0   
8959                                              Tiger       NaN     NaN   

      RRP (GBP)  RRP (USD)  RRP (CAD)  ...  Flag 3 not used  Flag 4 not used  \
0         29.99      34.99      39.99  ...              NaN              NaN   
1         19.99      19.99      24.99  ...              NaN              NaN   
2         19.99      19.99      24.99  ...              NaN              NaN   
3         14.99      14.99      17.99  ...              NaN              NaN   
4         29.99      34.99      39.99  ...              NaN              NaN   
...         ...        ...        ...  ...              ...              ...   
8955        NaN        NaN        NaN  ...              NaN              NaN   
8956        NaN        NaN        NaN  ...              NaN              NaN   
8957        NaN        NaN        NaN  ...              NaN              NaN   
8958        NaN        NaN        NaN  ...              NaN              NaN   
8959        NaN        NaN        NaN  ...              NaN              NaN   

      Flag 5 not used  Flag 6 not used  Flag 7 not used  Flag 8 not used  \
0                 NaN              NaN              NaN              NaN   
1                 NaN              NaN              NaN              NaN   
2                 NaN              NaN              NaN              NaN   
3                 NaN              NaN              NaN              NaN   
4                 NaN              NaN              NaN              NaN   
...               ...              ...              ...              ...   
8955              NaN              NaN              NaN              NaN   
8956              NaN              NaN              NaN              NaN   
8957              NaN              NaN              NaN              NaN   
8958              NaN              NaN              NaN              NaN   
8959              NaN              NaN              NaN              NaN   

      Value new (USD)  Value used (USD)  Launch date   Exit date  
0               32.44             16.33   01/01/2012  31/12/2013  
1               28.55              6.72   01/06/2012  31/12/2013  
2               50.00             11.19   01/01/2012  31/12/2013  
3               16.84              7.93   01/01/2012  31/07/2013  
4               60.00             19.06   01

Ebay data dataframe

In [342]:
df_ebay = pd.read_csv('lego_ebay_sold_price.csv', parse_dates=True)
df_ebay

title currency sold_price  \
0      Lego STAR WARS MILLENNIUM FALCON BRAND NEW 135...        £     108.79   
1      LEGO 40499 Santa's Sleigh (343 pcs) Brand New!...        £      67.73   
2      LEGO 76208 Marvel Thor Love & Thunder. The Goa...        £      37.49   
3      LEGO City 60161 - Jungle Exploration Site - BR...        £      145.0   
4      Lego 75299 Trouble on Tatooine new factory sea...        £       22.0   
...                                                  ...      ...        ...   
15732  Lego CHIMA 70147 : Sir Fangar's Ice Fortress N...        £     142.24   
15733  Lego Star Wars 9516 Jabba's Palace NEU/OVP/MIS...        £     345.96   
15734  Lego 71006 The Simpson House - FUORI PRODUZION...        £     338.16   
15735  Brand new LEGO Harry Potter: Hogwarts: Dumbled...        £      44.55   
15736         lego city high-speed passenger train 60051        £       57.0   

         sold_date  
0       9 Sep 2022  
1       9 Sep 2022  
2       9 Sep 2022  
3       9 Sep 2022  
4       9 Sep 2022  
...            ...  
15732  10 Sep 2022  
15733  10 Sep 2022  
15734  10 Sep 2022  
15735  10 Sep 2022  
15736  10 Sep 2022  

[15737 rows x 4 columns]

CLEAN EBAY DATA
Extract all numbers found in title into columns

In [343]:

# Extract all numbers in title to a list in a new DataFrame: normally LEGO's numbers code, number of pieces, set quantity, number of minifigures, etc.
df_numbers = df_ebay['title'].str.findall('([0-9]+)').to_list()

# Express all values in list to multi columns, each with one value.
# As with LEGO sets, the number of pieces is usually smaller than the actual number code. Other number are significantly smaller than the pieces and number code.
# Fill all nan value with 0, convert to int data type to find the max value

# df_numbers = df_ebay.numbers.apply(pd.Series).fillna(0).astype(int).max(axis=1)
df_numbers = pd.DataFrame([pd.Series(x, dtype = 'float') for x in df_numbers]).fillna(0).astype(int).max(axis=1) # Use this method to avoid the warning: 'The default dtype for empty Series will be 'object' instead of 'float64' in a future version warning'
df_numbers

0        75257
1        40499
2        76208
3        60161
4        75299
         ...  
15732    70147
15733     9516
15734    71006
15735    76402
15736    60051
Length: 15737, dtype: int64

In [344]:
# Update the column 'numbers' with the correct value
df_ebay.loc[:,'numbers'] = df_numbers
df_ebay

title currency sold_price  \
0      Lego STAR WARS MILLENNIUM FALCON BRAND NEW 135...        £     108.79   
1      LEGO 40499 Santa's Sleigh (343 pcs) Brand New!...        £      67.73   
2      LEGO 76208 Marvel Thor Love & Thunder. The Goa...        £      37.49   
3      LEGO City 60161 - Jungle Exploration Site - BR...        £      145.0   
4      Lego 75299 Trouble on Tatooine new factory sea...        £       22.0   
...                                                  ...      ...        ...   
15732  Lego CHIMA 70147 : Sir Fangar's Ice Fortress N...        £     142.24   
15733  Lego Star Wars 9516 Jabba's Palace NEU/OVP/MIS...        £     345.96   
15734  Lego 71006 The Simpson House - FUORI PRODUZION...        £     338.16   
15735  Brand new LEGO Harry Potter: Hogwarts: Dumbled...        £      44.55   
15736         lego city high-speed passenger train 60051        £       57.0   

         sold_date  numbers  
0       9 Sep 2022    75257  
1       9 Sep 2022    40499  
2       9 Sep 2022    76208  
3       9 Sep 2022    60161  
4       9 Sep 2022    75299  
...            ...      ...  
15732  10 Sep 2022    70147  
15733  10 Sep 2022     9516  
15734  10 Sep 2022    71006  
15735  10 Sep 2022    76402  
15736  10 Sep 2022    60051  

[15737 rows x 5 columns]

In [345]:
# Remove duplicates
df_ebay = df_ebay.drop_duplicates()
df_ebay

title currency sold_price  \
0      Lego STAR WARS MILLENNIUM FALCON BRAND NEW 135...        £     108.79   
1      LEGO 40499 Santa's Sleigh (343 pcs) Brand New!...        £      67.73   
2      LEGO 76208 Marvel Thor Love & Thunder. The Goa...        £      37.49   
3      LEGO City 60161 - Jungle Exploration Site - BR...        £      145.0   
4      Lego 75299 Trouble on Tatooine new factory sea...        £       22.0   
...                                                  ...      ...        ...   
15690  LEGO CITY 60285 Sports Car with minifigure - N...        £       9.75   
15694        LEGO 40515 Pirates and Treasure VIP polybag        £        6.0   
15705  Lego 60202 City Town People Pack - Outdoor Adv...        £       45.0   
15706  LEGO 40533 Ideas Cosmic Cardboard Adventures S...        £      17.99   
15717  LEGO 10280 Botanical Collection Flower Bouquet...        £      39.99   

         sold_date  numbers  
0       9 Sep 2022    75257  
1       9 Sep 2022    40499  
2       9 Sep 2022    76208  
3       9 Sep 2022    60161  
4       9 Sep 2022    75299  
...            ...      ...  
15690  10 Sep 2022    60285  
15694  10 Sep 2022    40515  
15705  10 Sep 2022    60202  
15706  10 Sep 2022    40533  
15717  10 Sep 2022    10280  

[11320 rows x 5 columns]

Are those any invalid values: set number smaller than 999 and larger than 99999

In [346]:
filter = ((df_ebay['numbers'] < 999) | (df_ebay['numbers'] > 99999))
filter.sum()

1799

There are 1799 invalid values. We're going to drop them.

In [347]:
# Take only row with a valid numbers: number set is from 999 to 99999. 
df_ebay = df_ebay[~filter]
df_ebay

title currency sold_price  \
0      Lego STAR WARS MILLENNIUM FALCON BRAND NEW 135...        £     108.79   
1      LEGO 40499 Santa's Sleigh (343 pcs) Brand New!...        £      67.73   
2      LEGO 76208 Marvel Thor Love & Thunder. The Goa...        £      37.49   
3      LEGO City 60161 - Jungle Exploration Site - BR...        £      145.0   
4      Lego 75299 Trouble on Tatooine new factory sea...        £       22.0   
...                                                  ...      ...        ...   
15690  LEGO CITY 60285 Sports Car with minifigure - N...        £       9.75   
15694        LEGO 40515 Pirates and Treasure VIP polybag        £        6.0   
15705  Lego 60202 City Town People Pack - Outdoor Adv...        £       45.0   
15706  LEGO 40533 Ideas Cosmic Cardboard Adventures S...        £      17.99   
15717  LEGO 10280 Botanical Collection Flower Bouquet...        £      39.99   

         sold_date  numbers  
0       9 Sep 2022    75257  
1       9 Sep 2022    40499  
2       9 Sep 2022    76208  
3       9 Sep 2022    60161  
4       9 Sep 2022    75299  
...            ...      ...  
15690  10 Sep 2022    60285  
15694  10 Sep 2022    40515  
15705  10 Sep 2022    60202  
15706  10 Sep 2022    40533  
15717  10 Sep 2022    10280  

[9521 rows x 5 columns]

In [348]:
df_ebay['numbers'].nunique()

3260

There are 3260 uniqe product code so we group them

In [349]:
df_ebay.groupby('numbers').value_counts()
# need to check how to count rows in one group


numbers  title                                                                             currency  sold_price  sold_date  
1017     Lego Figurine Minifig City homme man papa dad shirt chemise rayure red cty1017    £         2.42        8 Sep 2022     1
1050     Lego Technic Antriebsgummi für Motor 3x3 Blau                           (1050 #)  £         1.31        8 Sep 2022     1
1051     LEGO® Star Wars Minifigur Droide D-O sw1051 (NEU)                                 £         4.33        9 Sep 2022     1
1075     Lego Figur Zubehör Beine Hose Beige mit Dekor aus Star Wars             (1075 #)  £         1.31        8 Sep 2022     1
         Lego Figur Zubehör Oberteil Beige mit Dekor aus Star Wars               (1075 #)  £         1.31        8 Sep 2022     1
                                                                                                                               ..
98283    Lego Dark Tan 1x2 Wall effect Brick (98283) x10 in a set *NEW* City Star Wars     £   